File: packages/hero-parser/src-rs/mapping.rs
use std::collections::HashMap;
use std::sync::LazyLock;

pub struct HeroMapping {
    pub enum_key: &'static str,
    pub display_name: &'static str,
}

static HERO_MAP: LazyLock<HashMap<&'static str, HeroMapping>> = LazyLock::new(|| {
    let mut m = HashMap::new();

    let entries: Vec<(&str, &str, &str)> = vec![
        // heroes_wip
        //Deadlock internal name => Deadlock enum key => Deadlock display name
        ("abrams", "Abrams", "Abrams"),
        ("bookworm", "Paige", "Paige"),
        ("doorman", "Doorman", "Doorman"),
        ("doorman_v2", "Doorman", "Doorman"),
        ("drifter", "Drifter", "Drifter"),
        ("dynamo", "Dynamo", "Dynamo"),
        ("familiar", "Rem", "Rem"),
        ("fencer", "Apollo", "Apollo"),
        ("frank", "Victor", "Victor"),
        ("frank_v2", "Victor", "Victor"),
        ("geist", "LadyGeist", "Lady Geist"),
        ("haze", "Haze", "Haze"),
        ("inferno", "Infernus", "Infernus"),
        ("ivy", "Ivy", "Ivy"),
        ("lash", "Lash", "Lash"),
        ("mcginnis", "McGinnis", "McGinnis"),
        ("necro", "Graves", "Graves"),
        ("pocket", "Pocket", "Pocket"),
        ("priest", "Venator", "Venator"),
        ("punkgoat", "Billy", "Billy"),
        ("unicorn", "Celeste", "Celeste"),
        ("vampirebat", "Mina", "Mina"),
        ("werewolf", "Silver", "Silver"),
        ("wraith", "Wraith", "Wraith"),
        ("yamato", "Yamato", "Yamato"),
        // heroes_staging
        ("archer", "GreyTalon", "Grey Talon"),
        ("archer_v2", "GreyTalon", "Grey Talon"),
        ("astro", "Holliday", "Holliday"),
        ("atlas_detective", "Abrams", "Abrams"),
        ("atlas_detective_v2", "Abrams", "Abrams"),
        ("bebop", "Bebop", "Bebop"),
        ("butcher", "Wrecker", "Wrecker"),
        ("chrono", "Paradox", "Paradox"),
        ("digger", "MoKrill", "Mo & Krill"),
        ("engineer", "McGinnis", "McGinnis"),
        ("ghost", "LadyGeist", "Lady Geist"),
        ("gigawatt", "Seven", "Seven"),
        ("gigawatt_prisoner", "Seven", "Seven"),
        ("grey_talon", "GreyTalon", "Grey Talon"),
        ("haze_v2", "Haze", "Haze"),
        ("hornet_v3", "Vindicta", "Vindicta"),
        ("inferno_v4", "Infernus", "Infernus"),
        ("kelvin", "Kelvin", "Kelvin"),
        ("kelvin_explorer", "Kelvin", "Kelvin"),
        ("kelvin_v2", "Kelvin", "Kelvin"),
        ("lash_v2", "Lash", "Lash"),
        ("magician", "Sinclair", "Sinclair"),
        ("magician_v2", "Sinclair", "Sinclair"),
        ("mirage", "Mirage", "Mirage"),
        ("mirage_v2", "Mirage", "Mirage"),
        ("nano", "Calico", "Calico"),
        ("prof_dynamo", "Dynamo", "Dynamo"),
        ("shiv", "Shiv", "Shiv"),
        ("shiv_ult", "Shiv", "Shiv"),
        ("synth", "Pocket", "Pocket"),
        ("tengu", "Ivy", "Ivy"),
        ("vindicta", "Vindicta", "Vindicta"),
        ("viper", "Vyper", "Vyper"),
        ("viscous", "Viscous", "Viscous"),
        ("warden", "Warden", "Warden"),
        ("wraith_gen_man", "Wraith", "Wraith"),
        ("wraith_magician", "Wraith", "Wraith"),
        ("wraith_puppeteer", "Wraith", "Wraith"),
        ("wrecker", "Wrecker", "Wrecker"),
        ("yamato_v2", "Yamato", "Yamato"),
    ];

    for (internal, enum_key, display) in entries {
        m.insert(
            internal,
            HeroMapping {
                enum_key,
                display_name: display,
            },
        );
    }

    m
});

// Paths to look for heroes in
pub const HERO_PATH_PREFIXES: [&str; 3] = [
    "models/heroes_wip/",
    "models/heroes_staging/",
    "models/heroes/",
];

// Lookup hero by internal name
pub fn lookup_hero(internal_name: &str) -> Option<&'static HeroMapping> {
    HERO_MAP.get(internal_name)
}

File: packages/hero-parser/src/mapping.ts
type HeroMapping = {
  enumKey: string;
  displayName: string;
};

const HERO_MAP: Record<string, HeroMapping> = {
  // heroes_wip
  abrams: { enumKey: "Abrams", displayName: "Abrams" },
  bookworm: { enumKey: "Paige", displayName: "Paige" },
  doorman: { enumKey: "Doorman", displayName: "Doorman" },
  doorman_v2: { enumKey: "Doorman", displayName: "Doorman" },
  drifter: { enumKey: "Drifter", displayName: "Drifter" },
  dynamo: { enumKey: "Dynamo", displayName: "Dynamo" },
  familiar: { enumKey: "Rem", displayName: "Rem" },
  fencer: { enumKey: "Apollo", displayName: "Apollo" },
  frank: { enumKey: "Victor", displayName: "Victor" },
  frank_v2: { enumKey: "Victor", displayName: "Victor" },
  geist: { enumKey: "LadyGeist", displayName: "Lady Geist" },
  haze: { enumKey: "Haze", displayName: "Haze" },
  inferno: { enumKey: "Infernus", displayName: "Infernus" },
  ivy: { enumKey: "Ivy", displayName: "Ivy" },
  lash: { enumKey: "Lash", displayName: "Lash" },
  mcginnis: { enumKey: "McGinnis", displayName: "McGinnis" },
  necro: { enumKey: "Graves", displayName: "Graves" },
  pocket: { enumKey: "Pocket", displayName: "Pocket" },
  priest: { enumKey: "Venator", displayName: "Venator" },
  punkgoat: { enumKey: "Billy", displayName: "Billy" },
  unicorn: { enumKey: "Celeste", displayName: "Celeste" },
  vampirebat: { enumKey: "Mina", displayName: "Mina" },
  werewolf: { enumKey: "Silver", displayName: "Silver" },
  wraith: { enumKey: "Wraith", displayName: "Wraith" },
  yamato: { enumKey: "Yamato", displayName: "Yamato" },
  // heroes_staging
  archer: { enumKey: "GreyTalon", displayName: "Grey Talon" },
  archer_v2: { enumKey: "GreyTalon", displayName: "Grey Talon" },
  astro: { enumKey: "Holliday", displayName: "Holliday" },
  atlas_detective: { enumKey: "Abrams", displayName: "Abrams" },
  atlas_detective_v2: { enumKey: "Abrams", displayName: "Abrams" },
  bebop: { enumKey: "Bebop", displayName: "Bebop" },
  butcher: { enumKey: "Wrecker", displayName: "Wrecker" },
  chrono: { enumKey: "Paradox", displayName: "Paradox" },
  digger: { enumKey: "MoKrill", displayName: "Mo & Krill" },
  engineer: { enumKey: "McGinnis", displayName: "McGinnis" },
  ghost: { enumKey: "LadyGeist", displayName: "Lady Geist" },
  gigawatt: { enumKey: "Seven", displayName: "Seven" },
  gigawatt_prisoner: { enumKey: "Seven", displayName: "Seven" },
  grey_talon: { enumKey: "GreyTalon", displayName: "Grey Talon" },
  haze_v2: { enumKey: "Haze", displayName: "Haze" },
  hornet_v3: { enumKey: "Vindicta", displayName: "Vindicta" },
  inferno_v4: { enumKey: "Infernus", displayName: "Infernus" },
  kelvin: { enumKey: "Kelvin", displayName: "Kelvin" },
  kelvin_explorer: { enumKey: "Kelvin", displayName: "Kelvin" },
  kelvin_v2: { enumKey: "Kelvin", displayName: "Kelvin" },
  lash_v2: { enumKey: "Lash", displayName: "Lash" },
  magician: { enumKey: "Sinclair", displayName: "Sinclair" },
  magician_v2: { enumKey: "Sinclair", displayName: "Sinclair" },
  mirage: { enumKey: "Mirage", displayName: "Mirage" },
  mirage_v2: { enumKey: "Mirage", displayName: "Mirage" },
  nano: { enumKey: "Calico", displayName: "Calico" },
  prof_dynamo: { enumKey: "Dynamo", displayName: "Dynamo" },
  shiv: { enumKey: "Shiv", displayName: "Shiv" },
  shiv_ult: { enumKey: "Shiv", displayName: "Shiv" },
  synth: { enumKey: "Pocket", displayName: "Pocket" },
  tengu: { enumKey: "Ivy", displayName: "Ivy" },
  vindicta: { enumKey: "Vindicta", displayName: "Vindicta" },
  viper: { enumKey: "Vyper", displayName: "Vyper" },
  viscous: { enumKey: "Viscous", displayName: "Viscous" },
  warden: { enumKey: "Warden", displayName: "Warden" },
  wraith_gen_man: { enumKey: "Wraith", displayName: "Wraith" },
  wraith_magician: { enumKey: "Wraith", displayName: "Wraith" },
  wraith_puppeteer: { enumKey: "Wraith", displayName: "Wraith" },
  wrecker: { enumKey: "Wrecker", displayName: "Wrecker" },
  yamato_v2: { enumKey: "Yamato", displayName: "Yamato" },
};

const HERO_PATH_PREFIXES = [
  "models/heroes_wip/",
  "models/heroes_staging/",
  "models/heroes/",
];

export function lookupHero(internalName: string): HeroMapping | undefined {
  return HERO_MAP[internalName];
}

export function extractInternalName(path: string): string | undefined {
  for (const prefix of HERO_PATH_PREFIXES) {
    if (path.startsWith(prefix)) {
      const rest = path.slice(prefix.length);
      const segment = rest.split("/")[0];
      if (segment && segment.length > 0) {
        return segment.toLowerCase();
      }
    }
  }
  return undefined;
}
